### ISSUE TITLE
[Bug] ATC allow_mix_precision / allow_fp32_to_fp16 在 SAM3 vision encoder FPN neck 上造成灾难性精度崩溃 (与 torch_npu fp16 行为不一致)

### ISSUE THREAD

**Issue by ZhaoyuZhong0913** [Ascend/pytorch#125]
## 摘要 (TL;DR)

同一份 SAM3 (Meta, HF Transformers) vision encoder ONNX 文件:

- 用 `torch_npu 2.9.0` + `torch.float16` 在 910B3 上 **eager 推理** → 精度与 CPU fp32 **完全对齐**, top-10 命中率 100%
- 用 **ATC 8.5.0** 把同一份 ONNX 编成 OM, `--precision_mode=allow_mix_precision` 或 `allow_fp32_to_fp16` → top-10 命中率 **跌到 5–8%** (基本随机)
- 只有 `--precision_mode=must_keep_origin_dtype` (fp32) 才能保精度, 但延迟比 PyTorch fp16 还慢 12%

**期望**: ATC 的降精度路径至少应与 `torch_npu` 运行时的 fp16 行为等价. 当前两者行为不一致, 且没有中间档位可用.

## 环境

| 组件 | 版本 |
|---|---|
| 硬件 | Ascend 910B3 (`soc_version=Ascend910B3`) |
| OS | Ubuntu 22.04 aarch64 |
| NPU 驱动 | 24.1.0.3 |
| CANN / ATC | 8.5.0 |
| PyTorch | 2.9.0+cpu |
| torch_npu | 2.9.0 |
| transformers | 5.3.0 |
| 模型 | `facebook/sam3` (Sam3Model, ViT-L backbone + FPN neck + CLIP text + DETR detector) |

## 目标模型结构 (关键)

```
Vision encoder (问题定位在这里):
  ViT-L backbone (32 层, hidden_size=1024)
       │
       ▼
  Sam3VisionNeck (FPN):
    ConvTranspose2d 2x2 stride=2  →  fpn_feat_3 (H/32)
         │ + skip
         ▼
    ConvTranspose2d              →  fpn_feat_2 (H/16)   ← max_diff=1.46 ❌
         │ + skip
         ▼
    ConvTranspose2d              →  fpn_feat_1 (H/8)    ← max_diff=1.47 ❌
         │ + skip
         ▼
    ConvTranspose2d              →  fpn_feat_0 (H/4)    ← max_diff=2.7e-2 (自纠一些)
```

输入 `pixel_values [1,3,1008,1008]` fp32. FPN neck 只占 ~2% 参数 (7.8M / 454M), 但是精度崩溃的源头.

### 1) 导出 ONNX (fp32)

```python
import torch
from transformers import Sam3Model
model = Sam3Model.from_pretrained("facebook/sam3", torch_dtype=torch.float32).eval()

class VisionWrap(torch.nn.Module):
    def __init__(self, m):
        super().__init__(); self.m = m
    def forward(self, pixel_values):
        out = self.m.get_vision_features(pixel_values=pixel_values)
        return (out.last_hidden_state, *out.fpn_hidden_states, *out.fpn_position_encoding)

dummy = torch.randn(1, 3, 1008, 1008)
torch.onnx.export(
    VisionWrap(model), (dummy,), "sam3_vision_encoder.onnx",
    input_names=["pixel_values"],
    output_names=["last_hidden_state",
                  "fpn_feat_0","fpn_feat_1","fpn_feat_2","fpn_feat_3",
                  "fpn_pe_0","fpn_pe_1","fpn_pe_2","fpn_pe_3"],
    opset_version=17,
)
```

### 2) ATC 转 OM (3 种 precision_mode 对比)

```bash
source /usr/local/Ascend/ascend-toolkit/set_env.sh

for MODE in allow_mix_precision allow_fp32_to_fp16 must_keep_origin_dtype; do
  atc --framework=5 \
      --model=sam3_vision_encoder.onnx \
      --output=vision_${MODE} \
      --soc_version=Ascend910B3 \
      --input_format=NCHW \
      --input_shape="pixel_values:1,3,1008,1008" \
      --precision_mode=${MODE} \
      --log=error
done
```

### 3) 用 ONNX Runtime CPU 跑同一 ONNX 作为 ground truth, 对比 OM 各输出

```python
import numpy as np, onnxruntime as ort
# 用 pyACL 跑 OM, 得到 om_fpn_feat_0..3
# 用 ORT CPU 跑 ONNX, 得到 cpu_fpn_feat_0..3
for i in range(4):
    d = np.abs(om_fpn_feat[i] - cpu_fpn_feat[i])
    print(f"fpn_feat_{i}: max_diff={d.max():.3f} mean_diff={d.mean():.3e}")
```

### 各 precision_mode 下的端到端精度 (15 样本, 3 prompts/image)

| `precision_mode` | top-10 命中 | bbox IoU | mask IoU | vision p50 | 可用? |
|---|---|---|---|---|---|
| `allow_mix_precision` + `high_performance` | **0.08** | **0.027** | 0.15 | 190 ms | ❌ |
| `allow_fp32_to_fp16` | **0.05** | **0.019** | 0.07 | 240 ms | ❌ |
| `must_keep_origin_dtype` | **0.9933** | **0.946** | 0.92 | 317 ms | ✓ (但慢) |
| **torch_npu fp16 eager (对照组)** | **1.000** | **1.000** | 1.00 | 150 ms | ✓ |

### 各输出张量的 ORT-CPU vs OM 数值差异 (`allow_mix_precision` 模式)

| 输出 | shape | max_diff | mean_diff | 判定 |
|---|---|---|---|---|
| `fpn_pe_2` (位置编码) | `[1,256,72,72]` | 4.8e-7 | 4.9e-9 | ✓ |
| `fpn_feat_0` | `[1,256,288,288]` | 2.7e-2 | 1e-3 级 | ⚠ |
| **`fpn_feat_1`** | `[1,256,144,144]` | **1.47** | 5.2e-1 | ❌ |
| **`fpn_feat_2`** | `[1,256,72,72]` | **1.46** | 3.5e-1 | ❌ |
| `last_hidden_state` | `[1,5184,1024]` | 220+ | 6.1e-1 | ❌ |

fp16 的数值范围是 `[-65504, 65504]`, 但 `max_diff` 达到 **220** 说明中间层已经 overflow 到远超 fp16 可表示的幅度.

## 根因假设

1. **`ConvTranspose2d` 中间累加器似乎用了 fp16 存储**. FPN 上采样把低分辨率 feature 插值放大, 中间累加值幅度很容易超过 fp16 范围.
2. **Skip connection 的 `Add` 两端幅度差异大**. ViT backbone 累积 32 层残差后, 其输出的幅度远大于上采样分支, 两者 fp16 相加时小分支被 underflow 掉.
3. **ATC 粗暴把 fp32 cast 成 fp16**, 没有像 `torch_npu` 那样对 LayerNorm / Softmax / Div / 大动态范围 Add / ConvTranspose 等保留 fp32 累加.

## 为什么 PyTorch NPU fp16 能保精度

根据我们的观察, `torch_npu` 对数值敏感算子 (`LayerNorm`, `Softmax`, `F.scaled_dot_product_attention` 的部分分支, `torch.div`) 默认走**混合精度 kernel**: fp16 输入输出 + fp32 累加 + fp32 归一化. 这是 Ascend 工程师多年踩坑积累的隐式白名单.

ATC 的降精度路径似乎缺少这一层"兜底", 用户自己导 ONNX 送 ATC 就拿不到这份保障.

## 期望行为

任何满足以下之一都能解决:

1. **`allow_mix_precision` 默认对 `ConvTranspose2d` / 大动态范围 `Add` / `LayerNorm` 自动保 fp32 累加**, 对齐 `torch_npu` 的行为. (首选)
2. **提供 `--keep_dtype=<op_list.json>` 机制**, 用户可以手动指定某些 op 保 fp32. 当前 ATC 虽然已有类似机制 (`keep_dtype_config`), 但文档和样例对"怎么从 ONNX op name 映射到 keep 列表"说明不足, 且 PyTorch 导出的 op 名 (`/ve/neck/layers.X/conv_transpose`) 难以定位对应 PyTorch 源码中的具体 `nn.Module`.
3. **至少在文档里明确说明**: 对 ViT + FPN 类模型, `allow_mix_precision` 有风险, 推荐路径是 X.

## 关于 bf16 (尚未测)

我们没有单独跑 `allow_fp32_to_bf16` / `allow_mix_precision_bf16`. 但 bf16 尾数 7 bit < fp16 的 10 bit, 按相同的 FPN neck 数值累加链路分析, 预期精度**只会更差**. 如需 bf16 实测数据可补.

## Workaround 现状

目前业务只能继续用 `torch_npu` + `torch_dtype=float16` 在 eager 模式运行. 我们测过:
- OM + `must_keep_origin_dtype`: 精度达标但延迟 453 ms > PyTorch eager 的 400 ms, **反而更慢**, 失去 OM 优化意义
- OM + 低精度模式: 快但精度崩

## 愿意提供

- 完整 SAM3 ONNX (vision 1.8GB, head 1.5GB) 和 OM (3 种模式各一份)
- 精度对比脚本和原始 npz 数据
- 最小复现 K8s pod YAML
- 在你们指定的 CANN / torch_npu 版本上补测
- 补测 bf16 相关 precision_mode
